File: Coin Change.py
# https://leetcode.com/problems/coin-change/

import sys

class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        
        """
        OBJECTIVE: Return the fewest number of coins that you need to make up that amount. If that amount of money cannot be made up by any combination of the coins, return -1.
        
        Time Complexity: O((amount + 1) * coins). The outer-loop iterates amount + 1 times and each iteration involves iterating the coins list
        
        Space Complexity: O(n) where n = length of dp. The dp list's size will always be equal to amount + 1 because that's how the list was initialized
        """
        
        # Create a dp list where it'll hold N elements and each element has a maximum value of max int
        dp = [sys.maxsize] * (amount + 1)
        
        # Set the 0th element to 0. This is saying, "to reach 0 amount (the index), we only need 0 coins (the value at index)"
        dp[0] = 0
        
        # Iterate amount range
        # NOTE: Added +1 because range() is non-inclusive [)
        for curAmount in range(1, amount + 1):
            
            # Iterate coins list
            for coin in coins:
                
                # If remaining change greater than or equal to 0, save how many coins were needed to reach this amount
                if curAmount - coin >= 0:
                    
                    # Which option has less coins? The one we already have saved or that with 1 additional coin
                    # NOTE: Remember that the initial value is maximum int
                    dp[curAmount] = min(dp[curAmount], 1 + dp[curAmount])
                    
        # If index at amount isn't still at initial value, return value at index
        if dp[amount] != sys.maxsize:
            return dp[amount]
        
        # If function is still continuing, then the amount cannot be reached with any combination of coins
        return -1

File: Coin Change.java
// https://leetcode.com/problems/coin-change/

class Solution {

    public int coinChangeSub(int[] coins, int amount, int[] memo) {

        // If amount is less than 0, then exit because we used too many coins
        if (amount < 0) {return -1;}

        // If amount is 0, congratulations because we don't need anymore coins
        if (amount == 0) {return 0;}

        // If an optimal solution has already been calculated for "amount", then return it rather than recompute it.
        if (memo[amount - 1] != 0) {
            return memo[amount - 1];
        }

        // Iterate list of available coins for use
        int optimalSolution = Integer.MAX_VALUE;
        for (int coin: coins) {

            // Calculate the least amount of coins needed to reach "amount"
            int best_solution_for_coin = coinChangeSub(coins, amount - coin, memo);

            // If the least amount of coins is greater than 0 & is less than the previous' optimal solution, then update variable to our new optimal solution
            if (0 <= best_solution_for_coin && best_solution_for_coin < optimalSolution) {
                optimalSolution = best_solution_for_coin + 1;
            }
        }

        // If optimal solution is -1, then it's impossible to reach "amount" with our "coins"
        if (optimalSolution == Integer.MAX_VALUE) {
            memo[amount - 1] = -1;
        }
        else {
            memo[amount - 1] = optimalSolution;
        }

        return memo[amount - 1];
    }

    public int coinChange(int[] coins, int amount) {

        /*
         * Time Complexity: O(m * n) => O(m) where m = 1 to amount and n = # of coins. Because of recursion and memoization,
         * the runtime depends on m. There's a recursive call for coin in coins while amount doesn't equal or become less than 0. 
         * Memoization also helps to end a recursive call early.
         * 
         * Space Complexity: O(m * n) where m = 1 to amount and n = # of coins. Same reason as above. 
         * 
         * Reference: https://betterprogramming.pub/learn-dynamic-programming-the-coin-change-problem-22a104478f50 
         */

        if (amount < 0) {
            return -1;
        }
        else {
            return coinChangeSub(coins, amount, new int[amount + 1]);
        }
    }

    public int coinChange_solutionA(int[] coins, int amount) {
        
        // If amount is 0 and coins list isn't empty, return 0
        if (amount == 0 && coins.length > 0) {return 0;}
        
        // Check if amount already exist as a coin in coins array
        for (int coin: coins) {
            if (coin == amount) {return 1;}
        }
        
        // Create a dp array
        int[] dp = new int[amount + 1];
        Arrays.fill(dp, amount + 1);
        
        // Set 1st element as 0.
        // NOTE: You need 0 coins to reach 0 amount
        dp[0] = 0;
        
        // Find out how many coins are needed to reach i-value
        for (int i=1; i<amount+1; i++) {
            
            // Iterate coins list
            for (int coin: coins) {
                
                // If coin is greater than or equal to i-value, update the ith index in dp
                if (i - coin >= 0) {
                    
                    // Keep the minimum number of coins needed to reach i-value
                    dp[i] = Math.min(dp[i], dp[i - coin] + 1);
                }
            }
        }
        
        // If element doesn't equal to its original maximum value, return its new minimum value
        if (dp[amount] != amount + 1) {
            return dp[amount];
        }
        
        return -1;
    }
}
